RAD FW Release 1.5

ishanvermani opened the pull request:
- Divide by zero protection implemented for stepper frequency
- Modify PID control to have Max/Min threshold
- Added BangBang control for stepper control in PID driver
- Have PID Deadzone as EEPROM param adjustable by CAN

chosterto commented on firmware/RAD/release/Core/Src/main.c:
```diff
@@ -821,6 +833,33 @@ int main(void)
+                }
+                case SET_PID_MAX_OUTPUT:
+                {
+                    pid_1.Init.max_output_abs = (double) decode_uint16_big_endian(new_message->data);
```
Just a nitpick here but why does `max_output_abs` expect a `uint16` while `min_output_abs` is a `uint8` lol

Other than that, LGTM

chosterto reviewed (approved):
LGTM!

ishanvermani commented:
- Fixed bug that didn't convert angle output of bang bang control to stepper steps
- Cleaned up imports, bugs in uint16 communication types
- Converted floats to doubles for consistency
- Added the ability to set the offset of the stepper in the event of skewed limitswitch or wheel backlash

ishanvermani commented on firmware/RAD/release/Core/Src/main.c:
```diff
@@ -322,6 +349,9 @@ int main(void)
                         new_setpoint = max_angle;
                     }
+
+                    new_setpoint = new_setpoint;
```
this should be removed. Test

ishanvermani replied:
redundant. Removed

ishanvermani commented on firmware/RAD/release/Core/Src/main.c:
```diff
@@ -891,15 +974,30 @@ int main(void)
+                {
+                    PID_Update_RolloverCount(&pid_1);
+                }
+
```
Confirm operation of this

ishanvermani replied:
Fixed. Works. For Release 2.0, make Encoder/LS occur outside of State machine

ishanvermani commented on firmware/RAD/release/Core/Inc/main.h:
```diff
@@ -87,6 +87,9 @@ typedef struct __attribute__((packed)){
 	uint8_t DRVCTRL_INTPOL;// : 1;
 	uint8_t DRVCTRL_DEDGE;// : 1;
 	uint8_t DRVCTRL_MRES;// : 4;
+	uint16_t PID_ERROR_THRESHOLD;
```
change to PID_MIN_OUTPUT for consistency

ishanvermani replied:
Done

ishanvermani reviewed (commented):
3 things + Fix TMC communication.
